library/core/src/fmt/mod.rs

```diff
@@ -971,9 +971,14 @@ pub trait UpperHex {
 /// `p` formatting.
 ///
 /// The `Pointer` trait should format its output as a memory location. This is commonly presented
-/// as hexadecimal.
+/// as hexadecimal. For more information on formatters, see [the module-level documentation][module].
 ///
-/// For more information on formatters, see [the module-level documentation][module].
+/// Printing of pointers is not a reliable way to discover how Rust programs are implemented,
+/// because the act of reading an address of a value may change how the value is stored in memory,
+/// and prevent some code optimizations.
```

Comment:
"prevent some code optimizations", weirdly, is something people start relying on if you are unclear about it.

And we should be careful about "values stored in memory", because the lowered machine code must uphold the program as it is executed on the abstract machine, which has its own memory.

And sometimes these optimizations _aren't_ inhibited, but people file bugs because the result is confusing!